File: src/main/java/io/github/richardson/gameplay/Character.java
package io.github.richardson.gameplay;

import io.github.richardson.Key;
import io.github.richardson.menu.Settings;
import lombok.Getter;
import processing.core.PApplet;
import processing.core.PImage;
import processing.core.PVector;


public class Character {
    private final PApplet p;
    private ObstacleManager obstacles;
    private Key key;
    private Settings settings;
    private int radius, maxJump;
    @Getter
    private PVector pos;
    private float maxSpeed;
    private PVector vel = new PVector(0, 0);
    private float airInertia;
    private boolean canJump, right;
    private PImage sprite;

    public Character(PApplet p, ObstacleManager obstacleManager, Key key, Settings settings) {
        this.p = p;
        this.obstacles = obstacleManager;
        this.key = key;
        this.settings = settings;
        this.radius = settings.getPlayerRadius();
        this.maxJump = settings.getPlayerMaxJump();
        this.maxSpeed = settings.getPlayerMaxSpeed();
        this.airInertia = settings.getPlayerAirInertia();
        this.pos = settings.getPlayerStart();
        this.sprite = p.loadImage("textures/characterTexture.png");
    }

    public void movement() {    //TODO: time-based movement processing
        PVector max = obstacles.maxUntilCollide(pos, vel, radius).copy();
        if (max.mag() < vel.mag()) {
            if (vel.mag() > 0) {
                vel.setMag(max.mag() - settings.getEpsilon());
            } else {
                vel.setMag(max.mag() + settings.getEpsilon());
            }
        }
        pos.add(vel);
    }

    public void render() {
        p.pushMatrix();
        p.translate(pos.x, pos.y);
        if (!right) {
            p.scale(-1, -1);
        } else {
            p.scale(1, -1);
        }
        p.imageMode(PApplet.CENTER);
        p.image(sprite, -radius, radius, radius, -radius);
        p.popMatrix();
    }

    public void velocityUpdate() {
        if (key.isRight()) {
            if (canJump) {
                vel.x += maxSpeed;
            } else {
                vel.x += maxSpeed / airInertia;
            }
        }
        if (key.isLeft()) {
            if (canJump) {
                vel.x -= maxSpeed;
            } else {
                vel.x -= maxSpeed / airInertia;
            }
        }
        if (key.isJump() && canJump) {
            vel.y += maxJump;
            canJump = false;
        }
    }
}

File: src/main/kotlin/io/github/richardson/gameplay/Character.kt
package io.github.richardson.gameplay

import io.github.richardson.Key
import io.github.richardson.menu.Settings
import processing.core.PApplet
import processing.core.PImage
import processing.core.PVector

class Character(
    private val p: PApplet,
    private val obstacles: ObstacleManager,
    private val key: Key,
    private val settings: Settings
) {
    private val radius: Int
    private val maxJump: Int
    val pos: PVector
    private val maxSpeed: Float
    private val vel = PVector(0f, 0f)
    private val airInertia: Float
    private var canJump = true
    private val right = false
    private val sprite: PImage?

    init {
        this.radius = settings.playerRadius
        this.maxJump = settings.playerMaxJump
        this.maxSpeed = settings.playerMaxSpeed
        this.airInertia = settings.playerAirInertia
        this.pos = settings.playerStart
        this.sprite = p.loadImage("textures/characterTexture.png")
    }

    fun movement() {    //TODO: time-based movement processing
        val max = obstacles.maxUntilCollide(pos, vel, radius.toFloat()).copy()
        if (max.mag() < vel.mag()) {
            if (vel.mag() > 0) {
                vel.setMag(max.mag() - settings.epsilon)
            } else {
                vel.setMag(max.mag() + settings.epsilon)
            }
        }
        pos.add(vel)
    }

    fun render() {
        p.pushMatrix()
        p.translate(pos.x, pos.y)
        if (right) {
            p.scale(1f, -1f)
        } else {
            p.scale(-1f, -1f)
        }
        //p.imageMode(PApplet.CENTER);
        p.image(sprite, pos.x, pos.y, (radius * 2).toFloat(), (radius * 2).toFloat())
        p.popMatrix()
    }

    fun velocityUpdate() {
        if (key.isRight) {
            if (canJump) {
                vel.x += maxSpeed
            } else {
                vel.x += maxSpeed / airInertia
            }
        }
        if (key.isLeft) {
            if (canJump) {
                vel.x -= maxSpeed
            } else {
                vel.x -= maxSpeed / airInertia
            }
        }
        if (key.isJump && canJump) {
            vel.y += maxJump.toFloat()
            canJump = false
        }
    }
}
